xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/DefaultMentionsEventExecutor.java

```diff
@@ -0,0 +1,224 @@
+        for (int i = 0; i < nbThreads; i++) {
+            startConsumer();
+        }
+        JMXMentionsMBean mbean = new JMXMentions(this.queue, this::updateNbThreads, () -> this.consumers.size());
+        this.jmxRegistration.registerMBean(mbean, MBEAN_NAME);
+    }
+
+    private void startConsumer()
+    {
+        MentionsConsumer runnable = new MentionsConsumer();
+        this.consumers.add(runnable);
+        this.threadPoolProvider
+            .initializeThread(runnable)
+            .start();
+    }
+
+    @Override
+    public void dispose()
+    {
+        this.jmxRegistration.unregisterMBean(MBEAN_NAME);
+        for (MentionsConsumer consumer : this.consumers) {
+            consumer.halt();
+        }
+    }
+
+    /**
+     * Update the number of threads.
+     *
+     * If the number increases, starts new threads
+     * If the number decreases, removes existing threads until the new number is reached
+     * @param nbThread the new number of threads. Negative values are equivalent to 0.
+     */
+    public void updateNbThreads(int nbThread)
+    {
+        int currentNbThreads = this.consumers.size();
+        if (nbThread > currentNbThreads) {
+            for (int i = currentNbThreads; i < nbThread; i++) {
+                startConsumer();
+            }
+        } else {
+            for (int i = Math.max(nbThread, 0); i < currentNbThreads; i++) {
+                MentionsConsumer remove = this.consumers.remove(0);
+                remove.halt();
+            }
+        }
+    }
+
+    @Override
+    public void execute(DocumentReference documentReference, DocumentReference authorReference, String version)
+    {
+        try {
+            this.queue.put(new MentionsData()
+                               .setDocumentReference(documentReference.toString())
+                               .setAuthorReference(authorReference.toString())
+                               .setVersion(version)
+                               .setWikiId(documentReference.getWikiReference().getName())
+            );
+        } catch (InterruptedException e) {
+            this.logger
+                .warn("Error while adding a task to the mentions analysis queue. Cause [{}]", getRootCauseMessage(e));
+        }
+    }
+
+    @Override
+    public long getQueueSize()
+    {
+        return this.queue.size();
+    }
+
+    /**
+     * Consumer of the mentions analysis task queue.
+     */
+    public class MentionsConsumer implements Runnable
+    {
+        private boolean halt;
+
+        @Override
+        public void run()
+        {
+            while (!this.halt) {
+                consume();
+            }
+        }
+
+        /**
+         * Consume a single queue task.
+         */
+        public void consume()
+        {
+            MentionsData data = null;
+            try {
+                // slightly active wait in order to let the loop be stopped when halt is set to true, once every 
```

Comment:
An alternative is to put in the queue an elements used to both wake up the thread and order it to stop.

Reply:
In theory, by putting "stop" element in the queue for each consumer, they would all be stopped one by one.
The only issue is if the server halt in the middle of the stopping process. In this case some stop elements would be persisted on disk (since the queue is persistent). The works would stop their work when started again. But I guess it can be prevented by cleaning the queue of unwanted elements at startup.

Reply:
> In this case some stop elements would be persisted on disk (since the queue is persistent). The works would stop their work when started again.

Not if you use == on a static to check if a task is a STOP task (but you will still need to ignore "bad" tasks).